### PR TITLE
[terraform-resources] introduce parallelism

### PR DESCRIPTION
**Pull request by maorfr**

### CODE REVIEW

File: utils/openshift_resource.py [hunk 1]
```diff
@@ -209,19 +209,19 @@ def initialize_resource_type(self, cluster, namespace, resource_type):
...
-        if name in desired:
-            self._lock.release()
-            raise ResourceKeyExistsError(name)
-        desired[name] = value
-        self._lock.release()
+        with self._lock:
+            desired = \
+                self._clusters[cluster][namespace][resource_type]['desired']
+            if name in desired:
+                self._lock.release()
```
Comment (jmelis) [inline utils/openshift_resource.py]: I wonder if this is necessary. I would assume that once the raise is called, the lock will be automatically released as it exits the `with` context. I may be wrong though...

File: utils/terraform_client.py [hunk 2]
```diff
@@ -23,81 +26,144 @@ def __init__(self, body, integration, integration_version):
...
             tfs[name] = tf
         self.tfs = tfs
 
+    class initSpec(object):
+        def __init__(self, name, wd):
+            self.name = name
+            self.wd = wd
+
+    def init_init_specs(self, working_dirs):
+        init_specs = []
```
Comment (jmelis) [inline utils/terraform_client.py]: can you use a comprehension list here?

File: utils/terraform_client.py [hunk 3]
```diff
@@ -23,81 +26,144 @@ def __init__(self, body, integration, integration_version):
...
+
+    # terraform plan
     def plan(self, enable_deletion):
         errors = False
         deletions_detected = False
-        for name, tf in self.tfs.items():
-            return_code, stdout, stderr = tf.plan(detailed_exitcode=False)
-            error = self.check_output(name, return_code, stdout, stderr)
+
+        pool = ThreadPool(self.thread_pool_size)
```
Comment (jmelis) [inline utils/terraform_client.py]: we could probably define the pool as an instance variable and reuse it everytime: `self.pool`

File: utils/terraform_client.py [hunk 4]
```diff
@@ -23,81 +26,144 @@ def __init__(self, body, integration, integration_version):
...
-            tf = Terraform(working_dir=wd)
-            return_code, stdout, stderr = tf.init()
-            error = self.check_output(name, return_code, stdout, stderr)
-            if error:
-                return None
+        for name, tf in results:
             tfs[name] = tf
         self.tfs = tfs
 
+    class initSpec(object):
```
Comment (jmelis) [inline utils/terraform_client.py]: why class as opposed to a simple dict?

File: utils/terraform_client.py [hunk 5]
```diff
@@ -23,81 +26,144 @@ def __init__(self, body, integration, integration_version):
...
         return deletions_detected
 
+    # terraform apply
     def apply(self):
         errors = False
-        for name, tf in self.tfs.items():
-            return_code, stdout, stderr = tf.apply(auto_approve=True)
-            error = self.check_output(name, return_code, stdout, stderr)
+
+        pool = ThreadPool(1)  # TODO: replace 1 with self.thread_pool_size
```
Comment (jmelis) [inline utils/terraform_client.py]: TODO

File: utils/terrascript_client.py [hunk 6]
```diff
@@ -56,37 +59,80 @@ def __init__(self, integration, integration_prefix, oc_map):
...
 
     def populate(self, tf_query):
+        populate_specs = self.init_populate_specs(tf_query)
+
+        pool = ThreadPool(self.thread_pool_size)
+        pool.map(self.populate_tf_resources, populate_specs)
+
+        self.validate()
+
+    class populateSpec(object):
```
Comment (jmelis) [inline utils/terrascript_client.py]: there are several cases of snakeCase in this PR, do we have more of these? I don't like it very much

File: utils/terraform_client.py (hunk 5 above)
Comment (maorfr) [reply utils/terraform_client.py]: to remain TODO until we gain more experience :)

File: utils/openshift_resource.py (hunk 1 above)
Comment (maorfr) [reply utils/openshift_resource.py]: you are right, i just missed it. thanks!

File: utils/terraform_client.py (hunk 2 above)
Comment (maorfr) [reply utils/terraform_client.py]: done

File: utils/terraform_client.py (hunk 3 above)
Comment (maorfr) [reply utils/terraform_client.py]: done

File: utils/terraform_client.py (hunk 4 above)
Comment (maorfr) [reply utils/terraform_client.py]: we need a tuple-like iterable list

File: utils/terrascript_client.py (hunk 6 above)
Comment (maorfr) [reply utils/terrascript_client.py]: changed them to CamelCase

File: utils/terraform_client.py (hunk 4 above)
Comment (maorfr) [reply utils/terraform_client.py]: removed all relevant classes

File: utils/terrascript_client.py [hunk 7]
```diff
@@ -56,37 +59,72 @@ def __init__(self, integration, integration_prefix, oc_map):
...
 
     def populate(self, tf_query):
+        populate_specs = self.init_populate_specs(tf_query)
+
+        pool = ThreadPool(self.thread_pool_size)
+        pool.map(self.populate_tf_resources, populate_specs)
+
+        self.validate()
+
+    def init_populate_specs(self, tf_query):
```
Comment (jmelis) [inline utils/terrascript_client.py]: all these functions are better defined as list comprehensions

File: utils/terrascript_client.py [hunk 8]
```diff
@@ -56,37 +59,72 @@ def __init__(self, integration, integration_prefix, oc_map):
...
 
-    def get_vault_tf_secrets(self, type):
-        secrets = {}
-        for name, data in self.accounts:
-            secrets_path = data['secrets_path']
-            secret = vault_client.read_all(secrets_path + '/' + type)
-            secrets[name] = secret
-        return secrets
+    def init_vault_tf_secret_specs(self):
+        vault_specs = []
```
Comment (jmelis) [inline utils/terrascript_client.py]: list comprehension

File: utils/terrascript_client.py (hunk 7 above)
Comment (jmelis) [reply utils/terrascript_client.py]: on second thought these are complex functions. It's clear what they do, maybe a list comprehension is not necessary here.

**Pull request merged**